sdk/metric/sdk.go

```diff
@@ -115,16 +119,141 @@ type (
 		recorder export.Aggregator
 	}
 
+	observerResult struct {
+		observer *observer
+	}
+
+	int64ObserverResult struct {
+		result observerResult
+	}
+
+	float64ObserverResult struct {
+		result observerResult
+	}
+
+	observerCallback func(result observerResult)
+
+	observer struct {
+		meter      *SDK
+		descriptor *export.Descriptor
+		// recorders maps encoded labelset to the pair of
+		// labelset and recorder
+		recorders map[string]labeledRecorder
```

Comment:
This looks like it could be a source of unbounded memory growth. We need to be clear about when it is cleared.

Two possible opinions:
(1) We shouldn't need to store this map in the observer at all. We only need to de-duplicate a particular observation if another appearance of the same label set has been observed in the same interval. This means we only need such a map to be built up during collection, not persisted across collections. This would lead to new Aggregators being allocated for each observation.
(2) We should only need to store the map of observers that were used in the last collection interval, that way we can re-use their Aggregator object and avoid an allocation. However, we should be sure that if a label set stops being observed, the memory is cleared. However, we don't have an Aggregator API to "reset" at present, and I think that would be needed, since if we re-use an Aggregator across intervals we need a way to zero it out.

Reply:
As mentioned during the meeting - we remove an aggregator for a specific label set if the label set hasn't got any observation for two collection intervals in a row. If the aggregator map becomes empty, we drop the map too - see the [`checkpointObserver`](https://github.com/open-telemetry/opentelemetry-go/blob/0972daae77faa6bf6a37823fbca1422cbfdae8f6/sdk/metric/sdk.go#L590-L611) function. I admit that "two collection intervals" is arbitrary. Maybe we should simply state that if an aggregator/labelset didn't get any observation in the collection interval, it gets dropped. This is assuming that observation callbacks should be more or less a simple sequence of calls to `result.Observe(…)` (that is - no conditional observations).

Short notes about opinions above:
Ad (1): I can imagine having two observers with callbacks, where both are using the same label sets, and a custom aggregator selector that assigns minmaxsumcount aggregator to one observer and some last value stuff to the second observer. That's probably a fringe case, but in such a case we would have two different aggregators for the same label set, thus having a behavior depending on the order of iteration of the `observers` map.

Ad (2): That's quite similar to the current solution, but instead of a map of observers used in last collection interval we have a map (`map[encodedLabelSet]Aggregator`) in each observer. Entries are dropped from the map when no observation happens, so if the label set receives an observation again, we create an Aggregator again - this is a way of "zeroing" the aggregator.